### PR TITLE
Rewrite of GPIO TRD to follow the new traits.

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request updates the GPIO TRD to follow the current (2.0) traits. The TRD was never finalized, so we can just bump the revision number and move to finalization.

### Testing Strategy

NA

### TODO or Help Wanted

Review/reading.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Review by bradjc — DISMISSED** [review]

**Review by hudson-ayers — APPROVED** [review]

File: doc/reference/trd103-gpio.md [hunk 1]
```diff
@@ -30,147 +30,321 @@ individually since ports often do not group pins as they are actually used on a
...
-  * `kernel::hil::gpio::PinCtl`: Controlls the input mode on a single pin.
-  * `kernel::hil::gpio::Client`: handles the callback when a GPIO interrupt is
-    fired.
+The GPIO HIL is the kernel crate, in module hil::gpio. It provides the following traits:
+
+  * `kernel::hil::gpio::Output` controls an output pin.
+  * `kernel::hil::gpio::Input` controls an input pin.
+  * `kernel::hil::gpio::Configure` configures a pin.
+  * `kernel::hil::gpio::ConfigureInputOutput` configures a pin that can simultaneously be an
+     input and an output (some hardware supports this. It depends on `Configure`.
```
Comment (ppannuto) [inline doc/reference/trd103-gpio.md:39]:
```suggestion
     input and an output (some hardware supports this. It depends on `Configure`).
```

File: doc/reference/trd103-gpio.md [hunk 2]
```diff
@@ -30,147 +30,321 @@ individually since ports often do not group pins as they are actually used on a
...
-method. `identifier` MUST contain the value passed to the `Pin` trait's `enable_interrupts` method.
+`InterruptValueWrapper` implements `InterruptWithValue`, `Client`, 
+`Input`, `Output`, and `Configure`.
 
+6 Composite Traits: Pin, InterruptPin, InterruptValuePin
+========================================
 
-5 Example Implementation
+The GPIO HIL uses fine-grained traits in order to follow the security
+principle of least privilege. For example, that needs to be able to read
```
Comment (ppannuto) [inline doc/reference/trd103-gpio.md:318]:
```suggestion
principle of least privilege. For example, something that needs to be able to read 
```

File: doc/reference/trd103-gpio.md [hunk 3]
```diff
@@ -30,147 +30,321 @@ individually since ports often do not group pins as they are actually used on a
...
+pub trait Pin: Input + Output + Configure {}
+pub trait InterruptPin<'a>: Pin + Interrupt<'a> {}
+pub trait InterruptValuePin<'a>: Pin + InterruptWithValue<'a> {}
+```
+
+6 Example Implementation
 ========================================
 
-6 Authors' Address
+There are example implementations of the GPIO HIL for the Atmel
```
Comment (ppannuto) [inline doc/reference/trd103-gpio.md:332]:
```suggestion
As of this writing (April 2021; Tock v1.6 and v2.0), there are example implementations of the GPIO HIL for the Atmel
```

This list will inevitably fall out of date

**Review by ppannuto — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]

**Pull request merged**